File: src/data/spotter.ts
/**
 * Spotter — the practical histology exam turned into a race.
 *
 * A slide appears with one structure pinned but unnamed; the student names it
 * before their opponent does. Like Term Grid, a Spotter game is shared as a
 * link carrying a seed rather than a row in a table, so `buildSpotter` has to
 * be a pure function of its inputs — same slides and seed, same game, down to
 * the option order — with every draw coming from `seededRandom`. A stray
 * `Math.random()` here would give two people sitting the "same" spotter round
 * different structures with no error to notice.
 *
 * Pure module: no React, no storage, no clock.
 */

import { seededRandom, shuffle } from './seededRandom.ts'
import { objectivesOf, structuresAt, type HistologySlide, type Objective, type SlideStructure } from './histology.ts'

export interface SpotterRound {
  slideId: string
  slideTitle: string
  objective: Objective
  /** The structure being asked about — its pin is shown, its label is not. */
  structureId: string
  at: { x: number; y: number }
  /** The correct label plus distractors, already shuffled. */
  options: string[]
  answer: string
}

export interface SpotterGame {
  rounds: SpotterRound[]
  /** Why there is no game, when there are no rounds. */
  refusal: 'too_few_structures' | 'too_few_labels' | null
}

export const ROUNDS = 8
export const OPTIONS_PER_ROUND = 4
export const MIN_LABELS = OPTIONS_PER_ROUND // need enough distinct labels to fill one question

/** One structure, at one objective it is actually pinned at, on one slide. */
interface Candidate {
  slide: HistologySlide
  objective: Objective
  structure: SlideStructure
}

/**
 * Every askable (slide, objective, structure) triple.
 *
 * "Askable" means pinned at that objective specifically — `structuresAt`
 * is what keeps a structure pinned only at 4× from ever being asked at 40×,
 * since the pin the student would be shown simply does not exist there.
 */
function candidatesOf(slides: HistologySlide[]): Candidate[] {
  const candidates: Candidate[] = []
  for (const slide of slides) {
    for (const objective of objectivesOf(slide)) {
      for (const structure of structuresAt(slide, objective)) {
        candidates.push({ slide, objective, structure })
      }
    }
  }
  return candidates
}

/**
 * Distractor labels for one round, same-subject first.
 *
 * A structure never offers itself as a distractor for its own question — that
 * is enforced by label rather than id, so a structure elsewhere sharing the
 * exact same label as the answer is excluded too, which is what "never equal
 * to the answer" actually requires: two options reading identically would be
 * indistinguishable to the student regardless of which structure backs them.
 */
function distractorsFor(
  answer: SlideStructure,
  subjectId: string,
  slides: HistologySlide[],
  random: () => number,
): string[] {
  const sameSubject = new Set<string>()
  const otherSubject = new Set<string>()
  for (const slide of slides) {
    const bucket = slide.subjectId === subjectId ? sameSubject : otherSubject
    for (const structure of slide.structures) {
      if (structure.label === answer.label) continue
      bucket.add(structure.label)
    }
  }

  const needed = OPTIONS_PER_ROUND - 1
  const fromSameSubject = shuffle([...sameSubject], random).slice(0, needed)
  if (fromSameSubject.length >= needed) return fromSameSubject

  const stillNeeded = needed - fromSameSubject.length
  const fromElsewhere = shuffle([...otherSubject], random).slice(0, stillNeeded)
  return [...fromSameSubject, ...fromElsewhere]
}

/**
 * Build the game.
 *
 * @param slides  the published slides the game may draw structures from
 * @param seed    the number carried in the shared link; the whole game
 *                follows from it
 * @param rounds  ceiling on how many rounds the game holds. Fewer askable
 *                structures than this shortens the game rather than repeating
 *                one.
 */
export function buildSpotter(slides: HistologySlide[], seed: number, rounds: number = ROUNDS): SpotterGame {
  const candidates = candidatesOf(slides)
  // Nothing pinned anywhere means nothing can ever be asked, whatever the
  // label count says — checked before the label floor, not after.
  if (candidates.length === 0) return { rounds: [], refusal: 'too_few_structures' }

  const distinctLabels = new Set(slides.flatMap((slide) => slide.structures.map((s) => s.label)))
  // Below the floor, every round would either repeat an option or hand back
  // fewer than OPTIONS_PER_ROUND choices — a game that reads as broken rather
  // than merely short. Refuse instead of thinning the options.
  if (distinctLabels.size < MIN_LABELS) return { rounds: [], refusal: 'too_few_labels' }

  const random = seededRandom(seed)
  const order = shuffle(candidates, random)

  const askedStructureIds = new Set<string>()
  const built: SpotterRound[] = []

  for (const candidate of order) {
    if (built.length >= rounds) break
    // The same structure can be pinned at more than one objective, which would
    // otherwise offer it as two different candidates — collapse to one ask.
    if (askedStructureIds.has(candidate.structure.id)) continue
    askedStructureIds.add(candidate.structure.id)

    const distractors = distractorsFor(candidate.structure, candidate.slide.subjectId, slides, random)
    const options = shuffle([candidate.structure.label, ...distractors], random)

    built.push({
      slideId: candidate.slide.id,
      slideTitle: candidate.slide.title,
      objective: candidate.objective,
      structureId: candidate.structure.id,
      // structuresAt already guarantees this objective is present in `at`.
      at: candidate.structure.at[candidate.objective]!,
      options,
      answer: candidate.structure.label,
    })
  }

  return { rounds: built, refusal: null }
}

File: android/app/src/main/java/com/synapse/app/core/minigames/Spotter.kt
package com.synapse.app.core.minigames

import com.synapse.app.core.taxonomy.TaxonomyTerm

/**
 * Spotter, re-targeted at the medical glossary.
 *
 * Web's Spotter (`src/data/spotter.ts`) pins an unnamed structure on a live
 * histology slide and asks the student to name it from four options, drawing
 * on `useLiveHistology`. Android has no offline-synced histology slide media
 * (see the deferral note on `feature/minigames/MinigamesRepository.kt`), so
 * rather than fabricate histology content this port keeps the same *round*
 * shape web uses — the correct answer plus same-category-first shuffled
 * distractors, seed-deterministic — but "pins" a glossary term's definition
 * instead of a slide structure: the student reads the definition and spots
 * the term it names. No facts are generated; every prompt and option comes
 * straight from the published glossary.
 *
 * Pure module: no Android framework, no storage, no clock.
 */

enum class SpotterRefusal { TOO_FEW_TERMS }

data class SpotterRound(
    val termId: String,
    /** The definition "pinned" for this round — the term itself is what the student spots. */
    val prompt: String,
    val category: String,
    /** The correct term plus distractors, already shuffled. */
    val options: List<String>,
    val answer: String,
)

data class SpotterGame(val rounds: List<SpotterRound>, val refusal: SpotterRefusal?)

const val SPOTTER_ROUNDS = 8
const val SPOTTER_OPTIONS_PER_ROUND = 4

/** Need at least this many distinctly named terms to fill one round's options. */
const val SPOTTER_MIN_TERMS = SPOTTER_OPTIONS_PER_ROUND

/**
 * Distractor term names for one round, same-category first — mirrors web's
 * `distractorsFor` (same-subject-first) over categories instead of subjects.
 */
private fun distractorsFor(answer: TaxonomyTerm, terms: List<TaxonomyTerm>, random: () -> Double): List<String> {
    val sameCategory = LinkedHashSet<String>()
    val otherCategory = LinkedHashSet<String>()
    for (term in terms) {
        if (term.term == answer.term) continue
        if (term.category == answer.category) sameCategory += term.term else otherCategory += term.term
    }

    val needed = SPOTTER_OPTIONS_PER_ROUND - 1
    val fromSameCategory = shuffle(sameCategory.toList(), random).take(needed)
    if (fromSameCategory.size >= needed) return fromSameCategory

    val stillNeeded = needed - fromSameCategory.size
    val fromElsewhere = shuffle(otherCategory.toList(), random).take(stillNeeded)
    return fromSameCategory + fromElsewhere
}

/**
 * Build the game.
 *
 * @param terms the published glossary terms a round may be asked from
 * @param seed the seed carried in a shared link; the whole game follows from it
 * @param rounds ceiling on how many rounds the game holds. Fewer askable terms than this shortens the game rather than repeating one.
 */
fun buildSpotterGame(terms: List<TaxonomyTerm>, seed: Int, rounds: Int = SPOTTER_ROUNDS): SpotterGame {
    // Only a term with a definition can be "pinned" as a prompt.
    val askable = terms.filter { it.definition.isNotBlank() }
    val distinctNames = askable.map { it.term }.toSet()
    // Below the floor, every round would either repeat an option or hand back
    // fewer than SPOTTER_OPTIONS_PER_ROUND choices — refuse rather than thin
    // the options into something that reads as broken.
    if (distinctNames.size < SPOTTER_MIN_TERMS) return SpotterGame(emptyList(), SpotterRefusal.TOO_FEW_TERMS)

    val random = seededRandom(seed)
    val order = shuffle(askable, random)

    val built = mutableListOf<SpotterRound>()
    for (candidate in order) {
        if (built.size >= rounds) break
        val distractors = distractorsFor(candidate, askable, random)
        // Not enough distinctly named peers to fill this particular round's
        // options — skip it rather than under-fill.
        if (distractors.size < SPOTTER_OPTIONS_PER_ROUND - 1) continue
        val options = shuffle(listOf(candidate.term) + distractors, random)
        built += SpotterRound(
            termId = candidate.id,
            prompt = candidate.definition,
            category = candidate.category,
            options = options,
            answer = candidate.term,
        )
    }

    return SpotterGame(built, refusal = null)
}
